File: 0778-swim-in-rising-water/0778-swim-in-rising-water.py
import heapq

class Solution:
    def swimInWater(self, grid):
        n = len(grid)
        directions = [(1,0), (-1,0), (0,1), (0,-1)]
        visited = [[False]*n for _ in range(n)]
        
        # 優先隊列 (time, x, y)
        heap = [(grid[0][0], 0, 0)]
        visited[0][0] = True
        res = 0
        
        while heap:
            time, x, y = heapq.heappop(heap)
            res = max(res, time)
            if x == n-1 and y == n-1:
                return res
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny]:
                    visited[nx][ny] = True
                    heapq.heappush(heap, (grid[nx][ny], nx, ny))

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<bool>> visited(n, vector<bool>(n, false));
        // 最小堆: (time, x, y)
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        
        pq.push({grid[0][0], 0, 0});
        visited[0][0] = true;
        int res = 0;
        
        int dirs[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
        
        while (!pq.empty()) {
            auto cur = pq.top(); pq.pop();
            int time = cur[0], x = cur[1], y = cur[2];
            res = max(res, time);
            if (x == n-1 && y == n-1) return res;
            
            for (auto& d : dirs) {
                int nx = x + d[0], ny = y + d[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    pq.push({grid[nx][ny], nx, ny});
                }
            }
        }
        return res;
    }
};
